thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/client/ThirdEyeCacheRegistry.java

```diff
@@ -112,14 +112,27 @@ public void onRemoval(RemovalNotification<PinotQuery, ResultSetGroup> notificati
         LOGGER.info("Expired {}", notification.getKey().getPql());
       }
     };
-    // ResultSetGroup Cache
-    // TODO: have a limit on key size and maximum weight
-    LoadingCache<PinotQuery, ResultSetGroup> resultSetGroupCache =
-        CacheBuilder.newBuilder().removalListener(listener).expireAfterAccess(1, TimeUnit.HOURS)
-            .build(new ResultSetGroupCacheLoader(pinotThirdeyeClientConfig));
+
+    // ResultSetGroup Cache. The size of this cache is limited by the total number of buckets in all ResultSetGroup.
+    // We estimate that 1 bucket (including overhead) consumes 1KB and this cache is allowed to use up to 50% of max
+    // heap space.
+    LoadingCache<PinotQuery, ResultSetGroup> resultSetGroupCache = CacheBuilder.newBuilder()
+        .removalListener(listener)
+        .expireAfterAccess(1, TimeUnit.HOURS)
+        .maximumWeight(getApproximateMaxBucketNumber(50))
```

Comment:
create a constant for 50 and name the variable appropriately. Add a todo to make it configurable